component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,314 @@
-        String jti = claimsSet.getJWTID();
-        long currentTimeInMillis = System.currentTimeMillis();
-        long timeStampSkewMillis = OAuthServerConfiguration.getInstance().getTimeStampSkewInSeconds() * 1000;
-
-        if (isEmpty(jwtIssuer) || isEmpty(jwtSubject) || expirationTime == null || audience ==
-                null || (preventTokenReuse && jti == null)) {
-            return logAndReturnFalse("Mandatory fields(Issuer, Subject, Expiration time , " +
-                    "JWT ID or Audience) are empty in the given JSON Web Token.");
+    private boolean validateMandatoryFeilds(List<String> mandatoryClaims, ReadOnlyJWTClaimsSet claimsSet) throws OAuthClientAuthnException {
+
+        for (String mandatoryClaim : mandatoryClaims) {
+            if (claimsSet.getClaim(mandatoryClaim) == null) {
+                String errorMessage = "Mandatory field :" + mandatoryClaim + " is missing in the JWT assertion.";
+                if (log.isDebugEnabled()) {
+                    log.debug(errorMessage);
+                }
+                throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+            }
         }
+        return true;
+    }
+
+    // "REQUIRED. sub. This MUST contain the client_id of the OAuth Client."
+    public boolean validateSubject(String jwtSubject, String consumerKey) throws OAuthClientAuthnException {
+
+        return validateWithClientId(jwtSubject, consumerKey);
+    }
 
-        if (isNotEmpty(validIssuer) && !validIssuer.equals(jwtIssuer)) {
-            return logAndReturnFalse("Invalid Issuer:" + jwtIssuer + " in the given JSON Web Token.");
+    // "REQUIRED. iss. This MUST contain the client_id of the OAuth Client." when a valid issuer is not specified in the
+    // jwtValidator.
+    private boolean validateIssuer(String issuer, String consumerKey) throws OAuthClientAuthnException {
+
+        //check whether the issuer is client_id
+        if (isEmpty(validIssuer)) {
+            return validateWithClientId(issuer, consumerKey);
+        } else if (!validIssuer.equals(issuer)) {
+            String errorMessage = "Invalid value :" + issuer + " is found in the JWT. It should be equal to the: " +
+                    validIssuer;
+            if (log.isDebugEnabled()) {
+                log.debug(errorMessage);
+            }
+            throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
         }
+        return true;
+    }
 
-        if (!validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expirationTime.getTime(),
-                issuedAtTime.getTime())) {
-            return false;
+    private Boolean validateWithClientId(String jwtClaim, String consumerKey) throws OAuthClientAuthnException {
+
+        if (consumerKey == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Unable to find OAuth application for provided JWT claim: " + jwtClaim);
```

Comment:
when consumerkey is null, saying ""Unable to find OAuth application for provided JWT claim: " + jwtClaim" doesn't look appplicable. Let's simplify this method offline.